cinder/volume/drivers/vmware/vmdk.py

```diff
@@ -2236,10 +2369,166 @@ def do_setup(self, context):
 
         self.volumeops.build_backing_ref_cache()
 
+        # Cache the storage profiles, so we don't
+        # have to fetch them every time.
+        if self.configuration.vmware_storage_profile:
+            self._get_storage_profiles()
+
         LOG.info("Successfully setup driver: %(driver)s for server: "
                  "%(ip)s.", {'driver': self.__class__.__name__,
                              'ip': self.configuration.vmware_host_ip})
 
+    def _get_storage_profiles(self):
+        """Fetch the list of configured storage profiles we use."""
+
+        LOG.debug("Profiles = '%s'", self.configuration.vmware_storage_profile)
+        for profile in pbm.get_all_profiles(self.session):
+            if profile.name in self.configuration.vmware_storage_profile:
+                profile_dict = {"name": profile.name,
+                                "id": profile.profileId}
+                self._storage_profiles.append(profile_dict)
+
+    def _volume_provider_metadata(self, volume, backing=None):
+        if not backing:
+            backing = self.volumeops.get_backing(volume.name, volume.id)
+
+        ds = self.volumeops.get_datastore(backing)
+        summary = self.volumeops.get_summary(ds)
+        profile = self._get_storage_profile(volume)
+        vcenter_uuid = (
+            self.session.vim.service_content.about.instanceUuid
+        )
+        provider_info = {
+            'vmware_vcenter_id': vcenter_uuid,
+            'vmware_ds_name': summary.name,
+            'vmware_profile_name': profile,
+        }
+
+        return provider_info
+
+    @utils.trace
+    def _get_datastores_for_profiles(self):
+        datastores = {}
+        for profile in self._storage_profiles:
+            for h in pbm.filter_hubs_by_profile(self.session,
+                                                None,
+                                                profile['id']):
+                if h.hubType != "Datastore":
+                    # We are not interested in Datastore Clusters for now
+                    continue
+                if h.hubId not in datastores:
+                    # Reconstruct a managed object reference to that
+                    # datastore
+                    ds = vim_util.get_moref(h.hubId, "Datastore")
+                    summary = self.volumeops.get_summary(ds)
+                    datastores[summary.name] = {'summary': summary,
+                                                'storage_profile': profile}
+
+        return datastores
+
+    def _new_host_for_volume(self, volume):
```

Comment:
I don't really like this function. It generates a cinder_host (you renamed in a lot of places), but only if the pool extracted from the current cinder_host doesn't match the backing's datastore.

I suggest to rename it to `generate_cinder_host` and let it always generate the host for the volume and do the check for matching outside in that one place. The other place calling this function doesn't handle a `None` very good anyways.